Add DTLS_get_data_mtu()

dwmw2 opened the pull request:
Best case, this is all relatively sane and I just need some help finishing `SSL_CIPHER_get_overhead()` for non-AEAD, non-CBC modes, and perhaps actually report errors instead of just returning zero if something goes wrong.

The idea is that `SSL_CIPHER_get_overhead()` really does just describe the cipher, in isolation — which requires three values: the external overhead, the block size that it's rounded up to, and the internal overhead before that rounding.

Then we have `DTLS_get_data_mtu()` which actually does that, as well as subtracting  `DTLS1_RT_HEADER_LENGTH`, from `s->d1->mtu`.

davidben commented on include/openssl/ssl.h:
```diff
@@ -1315,6 +1318,10 @@ __owur const SSL_CIPHER *SSL_get_current_cipher(const SSL *s);
+__owur int SSL_CIPHER_get_overhead(const SSL_CIPHER *c,
+				   unsigned int *int_overhead,
+				   unsigned int *blocksize,
+				   unsigned int *ext_overhead);
```
I think it's better to make this a function of the SSL rather than the SSL_CIPHER. A cipher's overhead varies by TLS version. Block ciphers in TLS 1.1 have an explicit IV while TLS 1.0 does not. AES-GCM in TLS 1.3 will have a different construction from TLS 1.2 and lose the explicit nonce.

You might also consider reporting a "max overhead" rather than try to slice it up this way. It's less precise for the legacy CBC-mode ciphers, but a simpler interface. I dunno if we'd ever gain a variable-length AEAD in the future (would require TLS 1.3 since 1.2's AD construction doesn't allow it), but that would also generalize better if we ever did. If you prefer to, it also lets you include the record header and stuff like TLS 1.3's slightly funny encrypted content type byte which affects these computations, rather than make callers be aware of all that.

dwmw2 commented:
@davidben thanks; that's useful feedback. However, note that this is only for DTLS — where we *do* have an explicit IV in all cases, don't we?

We do expose it via `DTLS_get_data_mtu()` which can make TLS-version-specific adustments (just as it's currently deducting `DTLS1_RT_HEADER_LENGTH`. I was still aiming for having the `SSL_CIPHER`-specific knowledge come from the `SSL_CIPHER` though. I'll take a closer look at DTLS1.3...

dwmw2 commented:
> . AES-GCM in TLS 1.3 will have a different construction from TLS 1.2 and lose the explicit nonce.

CCM won't have an explicit nonce either, will it? And ChaCha20 never did. That's all there is in TLSv1.3 for now — and it can't inherit additional ciphersuites from TLSv1.2 because they'll need to be  re-specified for TLSv1.3 anyway, right?

So *nothing* has an explicit nonce in TLSv1.3. But this is a hypothetical conversation... OpenSSL doesn't support the TLSv1.3 draft yet. I'm not going to lose *too* much sleep over it. So maybe I can split out the IV part of the "`ext_overhead`"  field so that `DTLS_get_data_mtu()` can decide whether to subtract it or not...  or maybe the `SSL_CIPHER` is going to have to change more dramatically for TLSv1.3 anyway.

I think for now it's sufficient just to avoid painting ourselves into a corner. I'll stop exporting `SSL_CIPHER_get_overhead()` and then we are left with just `DTLS_get_data_mtu()`. Which is indeed *"a function of the `SSL` not the `SSL_CIPHER`"* as you suggested. Sure, its internal implementation only works for TLS <= 1.2 right now, but that's OK.

dwmw2 commented:
I can disable the `->min_dtls` check in the unit test and it works for every ciphersuite except GOST, FWIW. At least, it returns a set of values... I haven't yet established that they are all correct!

Do we have an existing test which actually *establishes* DTLS connections with every supported ciphersuite? I could add a call to `DTLS_set_data_mtu()` to that... for added fun I could even try sending a packet of the maximum indicated data MTU, and check that it doesn't exceed the SSL MTU. And then try sending a payload one byte larger, and check that it *does* exceed the SSL MTU (or just fail to send). Is it worth doing those?

kaduk commented:
>  Is it worth doing those?

It seems like that is exactly the sort of thing that is worth testing!  (My non-team member two cents.)

mattcaswell commented:
Sorry, not had time to properly look at this (I'm at the OpenSSL face-to-face all week)...but...

> Do we have an existing test which actually establishes DTLS connections with every supported ciphersuite? 

No, I don't think we do, although we do have one that does this for TLS (see test/recipes/80-test_ssl_old.t around line 468). It would be great to extend it to run through the list again for DTLS.

>  Is it worth doing those?

A big YES from me!

dwmw2 commented:
This yak is still fairly hirsute

dwmw2 commented:
Now it at least checks that my `ssl_cipher_get_overhead()` function returns *a* value for each possible ciphersuite. It really ought to actually check that they're *correct* though. And especially for CBC ciphersuites it wants to check that they're correct for a range of MTU values too. That might end up being a completely new test.

dwmw2 commented:
Now with added test cases.

For each PSK ciphersuite *(which should cover every enc+mac combination;. no need to duplicate the tests for all auth/kx variants)* we call `SSL_set_mtu()` with each record MTU between 500 and 529, noting the payload MTU that `DTLS_get_data_mtu()` returns in each case.

Then we set the MTU up high so it's not relevant, and we send a packet of each payload size from the smallest to the largest. We note the actual record size which results from each payload.

Now it's just a simple case of checking if we were lied to — for each `SSL_set_mtu(r); p=DTLS_get_data_mtu();` pair, if we see that sending a payload of size `p` will result in a record larger than `r`. **OR** if we see that sending a payload larger than `p` results in a record which is not larger than `r`, we fail the test.

We disabled SSL compression, right? :)

Is there a way to disable encrypt_then_mac?

dwmw2 commented:
Hm... when encrypt_then_mac is disabled because I'm actually using this "in anger" from OpenConnect, I get the right answers. When I manually comment out the bit in `ssl_add_clienthello_tlsext()` which sends the extension, the payload MTU gets calculated accordingly but then the test fails because it seems to be still sending packets the same size as it would *with* encrypt-then-mac. So I suspect the fault there is in my hackish method of disabling encrypt-then-mac.... but I can't see what it is. More clueful eyes on it would be appreciated...

dwmw2 commented:
I think I have it working now; both for encrypt-then-mac, and mac-then-encrypt. Not that I can see how we can ever enable the former again.

dwmw2 commented:
The first two commits are already submitted in #1705 — fixing Encrypt-then-MAC and making it optional was kind of necessary in order to implement the test cases which need to explicitly test the result of `DTLS_get_data_mtu()` both with and without it.

Now I have the explicit `dtls_mtu_test` I've dropped the commits which added DTLSv1.0 and DTLSv1.2 to the cipher suite tests in `ssl_test_old`. I've moved them to #1710.

mattcaswell commented on ssl/d1_lib.c:
```diff
@@ -1088,3 +1091,39 @@ unsigned int dtls1_min_mtu(SSL *s)
+
+	/* Round encrypted payload down to cipher block size (for CBC etc.) */
+	if (blocksize)
+		mtu -= (mtu % blocksize);
```
Similarly here, you are subtracting from `mtu`...couldn't we go -ve?

dwmw2 replied:
MTU % *anything* cannot exceed MTU

mattcaswell commented on ssl/ssl_ciph.c:
```diff
@@ -1945,3 +1945,55 @@ int SSL_CIPHER_is_aead(const SSL_CIPHER *c)
+	if (c->algorithm_enc & SSL_AESGCM) {
+		out = EVP_GCM_TLS_EXPLICIT_IV_LEN + EVP_GCM_TLS_TAG_LEN;
+	} else if (c->algorithm_enc & (SSL_AES128CCM | SSL_AES256CCM)) {
+		out = EVP_CCM_TLS_EXPLICIT_IV_LEN + 16;
```
What is the 16 here? If no `'#define` for it, can we at least add a comment?

dwmw2 replied:
I couldn't find a define for the CCM and ChaCha20/Poly1305 overheads. In context I wasn't sure what value a comment would add...

mattcaswell commented on ssl/ssl_ciph.c:
```diff
@@ -1945,3 +1945,55 @@ int SSL_CIPHER_is_aead(const SSL_CIPHER *c)
+	} else if (c->algorithm_enc & (SSL_AES128CCM | SSL_AES256CCM)) {
+		out = EVP_CCM_TLS_EXPLICIT_IV_LEN + 16;
+	} else if (c->algorithm_enc & (SSL_AES128CCM8 | SSL_AES256CCM8)) {
+		out = EVP_CCM_TLS_EXPLICIT_IV_LEN + 8;
```
Same here

mattcaswell commented on ssl/ssl_ciph.c:
```diff
@@ -1945,3 +1945,55 @@ int SSL_CIPHER_is_aead(const SSL_CIPHER *c)
+	} else if (c->algorithm_enc & (SSL_AES128CCM8 | SSL_AES256CCM8)) {
+		out = EVP_CCM_TLS_EXPLICIT_IV_LEN + 8;
+	} else if (c->algorithm_enc & SSL_CHACHA20POLY1305) {
+		out = 16;
```
And here

dwmw2 replied:
This is also covered by a comment added further up. If I explicitly reply to these, does github stop showing that there are outstanding review comments? ...

kaduk replied:
I think it uses a heuristic based on whether the line being commented upon has changed.

mattcaswell commented on test/dtls_mtu_test.c:
```diff
@@ -0,0 +1,213 @@
+		const SSL_CIPHER *cipher = sk_SSL_CIPHER_value(ciphers, i);
+		const char *cipher_name = SSL_CIPHER_get_name(cipher);
+
+		if (strncmp(cipher_name, "PSK-", 4))
```
Why are we skipping these?

dwmw2 replied:
No point in testing the same suite twice with DH and without...

mattcaswell commented on test/dtls_mtu_test.c:
```diff
@@ -0,0 +1,213 @@
+
+	}
+	if (debug)
+		printf("Channel established\n");
```
I think all of the above boilerplate can be replaced with calls to some of the functions in ssltestlib.c, e.g. call create_ssl_ctx_pair(), create_ssl_objects() and create_ssl_connection(). See execute_test_large_message() in sslapittest.c for an example of usage.

mattcaswell commented on test/dtls_mtu_test.c:
```diff
@@ -0,0 +1,213 @@
+			printf("record %d for payload %d\n", reclen, i);
+
+		for (j = 0; j < 30; j++) {
+			if (i <= mtus[j] && reclen > (500 + j)) {
```
Shouldn't `reclen == 500 +j` where `i == mtus[j]`?

dwmw2 replied:
Not with CBC. For example, a record size (`SSL_set_mtu()`) of anything between (j+500=) 503-519 might *all* give a data MTU of (i=) 477. 

I'm fairly sure the two tests here *are* what we need to check for. There was plenty of caffeine involved, as well as empirical testing and even some bug fixes which I silently reworked into the `DTLS_get_data_mtu()` implementation.

They test that sending a given payload of size `i` will never result in a DTLS record larger than **any** of the record MTU values for which a max payload size <= `i` was reported, and that a given payload of size `i` will never result in a DTLS record which is *smaller* than any of the record MTU values for which a max payload size > `i` was reported.

The former for obvious reasons, the latter because it would mean that `DTLS_get_data_mtu()` is reporting a value which is too low -- it *does* fit in the requested record MTU, but *more* could fit too.

dwmw2 commented on ssl/d1_lib.c:
```diff
@@ -1088,3 +1091,39 @@ unsigned int dtls1_min_mtu(SSL *s)
+	if (ext_overhead >= mtu)
+		return 0;
+
+	mtu -= ext_overhead + DTLS1_RT_HEADER_LENGTH;
```
Good spot.  The testing showed me I forget the header but then in fixing it I forgot the guard. Will fix.

dwmw2 commented:
Fixed the overflow with message header, some updated comments. Using `create_ssl_objects()` and `create_ssl_connection()` from `ssltestlib.c` but not `create_ssl_ctx_pair()` since there's not actually a lot of overlap between what that does, and what I'm doing.

Attribution still to David.Woodhouse@intel.com — if it were two separate people, then it would not be correct for me (@infradead.org) to claim authorship just because I've made a few minor tweaks. It was the Intel guy who did all the hard work. The CLA thing is sorted anyway, right? There's no actual need to remove my @intel.com address from the list until any pending submissions are merged.

dwmw2 commented:
FWIW I pondered a macro for the `if (mtu < x) return 0 else mtu -= x` thing but figured it was just obfuscation, for two uses of it. Now the initial moronic bugs are out of the way it's not like these are likely to change ever again.

mattcaswell commented on ssl/d1_lib.c:
```diff
@@ -233,6 +234,8 @@ long dtls1_ctrl(SSL *s, int cmd, long larg, void *parg)
         s->d1->mtu = larg;
         return larg;
+    case DTLS_CTRL_GET_DATA_MTU:
+        return (long)dtls1_get_data_mtu(s);
```
This is why I hate SSL_ctrl for things like this (at some point I'd like to convert lots of the existing SSL_ctrls to be functions). The cast to long is just wrong and horrible. I'm wondering whether a proper function rather than a ctrl would be better?

mattcaswell commented on ssl/d1_lib.c:
```diff
@@ -1088,3 +1091,39 @@ unsigned int dtls1_min_mtu(SSL *s)
+{
+	unsigned int mac_overhead, int_overhead, blocksize, ext_overhead;
+	const SSL_CIPHER *ciph = SSL_get_current_cipher(s);
+	unsigned mtu = s->d1->mtu;
```
I am currently working on a size_t-ify libssl branch which converts s->d1->mtu to size_t. Should we anticipate that branch and just make it a size_t here and then (if we make this a proper function rather than a ctrl):

````
size_t DTLS_get_data_mtu(SSL *s);
````

I know that is inconsistent with the other mtu getting functions....but still...

dwmw2 commented:
Pushed the trivial rebase. Regarding making it a proper function and returning `size_t`.... I'm happy to do that, or I'm happy to leave it as-is, and let you sort it out when you rebase your size_t-ify branch on top of it and (perhaps?) turn the other MTU calls into proper functions too. I agree that it all needs doing, but don't care much about which order it's done it. So don't ask; just *tell* me to refactor it into a proper function and I'll do so :)

mattcaswell commented:
The size_t-ify branch (see #1751) touches pretty much the entire of libssl. If we don't do it now it may turn into a forgotten back-water of libssl that we forgot to convert. Let's convert it into a proper function now and make it a size_t. We won't convert the other functions yet (one step at a time).

dwmw2 commented on util/libssl.num:
```diff
@@ -403,3 +403,4 @@ SSL_dane_clear_flags                    403	1_1_0	EXIST::FUNCTION:
 SSL_SESSION_get0_cipher                 404	1_1_0	EXIST::FUNCTION:
 SSL_SESSION_get0_id_context             405	1_1_0	EXIST::FUNCTION:
 SSL_SESSION_set1_id                     406	1_1_0	EXIST::FUNCTION:
+DTLS_get_data_mtu                       407	1_1_1   EXIST::FUNCTION:DTLS
```
Is this correct? I note no *other* functions appear to depend on DTLS; not even `DTLS_method()`...

dwmw2 replied:
Que? I added that to libssl.num manually... should I not have done?

mattcaswell replied:
Oh! No - don't do that. Run "make update"

dwmw2 commented:
I left the internal `ssl_cipher_get_overhead()` function still using `unsigned int`. The values there are always going to be small, and an int seemed reasonable. Looking at the extent of the changes in #1751 ... should I change that too?

mattcaswell commented:
They are sizes so they should be size_t. Otherwise there are lots of implicit casts between "unsigned int" and size_t. Not actually a problem in this case (I don't think) but often they are (especially when you go the other way, i.e. size_t -> unsigned int). It seems unnecessary to use unsigned int, when we're going to end up converting the result to a size_t anyway.

dwmw2 commented:
What is your strategy for printing `size_t` values from unit tests, given that Windows lacks runtime support for the `%zu` format modifiers? I made the cipher overhead test just *refrain* from printing the values if `_WIN32` is defined, since it's purely cosmetic. Baulked at doing the same for `dtls_mtu_test` and left it using `int` for the moment...

dwmw2 commented:
So using `size_t` leads to build failures even with `%zu` on non-Windows platforms...

kroeckx commented:
Since we're actually still targeting C90, "%zu" is not supported. We should probably use a macro like PRIu32 and map it to the correct format specifier for those platforms that don't support %z.

kroeckx commented on ssl/d1_lib.c:
```diff
@@ -1088,3 +1088,39 @@ unsigned int dtls1_min_mtu(SSL *s)
     return dtls1_link_min_mtu() - BIO_dgram_get_mtu_overhead(SSL_get_wbio(s));
 }
+
+size_t DTLS_get_data_mtu(const SSL *s)
```
Please document this new public function.

dwmw2 commented:
> Since we're actually still targeting C90, "%zu" is not supported. We should probably use a macro like PRIu32 and map it to the correct format specifier for those platforms that don't support %z.

Perhaps I can retarget this to land before #1751 and we can resolve that there where it belongs?

kroeckx commented:
It's not actually using %zu anywhere, but it was at least not using it where it should so I've added that there.

dwmw2 commented:
OK, if you're going to make me care about `size_t` *(and I might as well; I've spent enough time shaving the yak @nmav gave me that this is fairly much in the noise ☺)* then here it is with a `PRIzu` macro to replace `%zu` for prehistoric unmaintained platforms, and a man page for `DTLS_get_data_mtu()`. There's nothing to be done to add the man page to an index or a makefile somewhere?

mattcaswell commented:
Oh. I didn't see your PRIzu commit until after I did commit ebd852b1eac7 in #1751

dwmw2 commented:
Rebased on top of `OSSLzu` from #1787

dwmw2 commented:
Don't think those CMS test failures are mine? 
```
wine ../apps/openssl.exe cms -EncryptedData_encrypt -in ../test/smcont.txt -outform PEM -des3 -secretkey 000102030405060708090A0B0C0D0E0F1011121314151617 -stream -out test.cms => 139
    not ok 26 - encrypted content test streaming PEM format, triple DES key
    
    #   Failed test 'encrypted content test streaming PEM format, triple DES key'
    #   at ../test/recipes/80-test_cms.t line 429.
```

levitte commented:
exit code 139 is SIGSEGV, FYI...  reference to NULL, perhaps?

dwmw2 commented:
Right, but I don't think I've changed any code that's even *run* in that test, have I?

levitte commented on test/recipes/70-test_dtls_mtu.t:
```diff
@@ -0,0 +1,21 @@
+#! /usr/bin/env perl
```
FYI, I feel unsure this is a test of the PACKET layer.  See `test/README` for the numeric groups.

levitte commented:
Re the SIGSEGV, you're right, I don't see how you could have caused that in that case.  Are you rebased on top of a fresh master?

dwmw2 commented:
Test numbering fixed. Yes, this is on top of a fresh master — it needed to be, for #1787.

dwmw2 commented:
FWIW my own master build with mingw64 fails differently: 
```
Test Summary Report
-------------------
../test/recipes/30-test_evp.t            (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=90, Tests=432, 326 wallclock secs ( 1.56 usr  0.21 sys + 149.71 cusr 154.03 csys = 305.51 CPU)
Result: FAIL
Failed 1/90 test programs. 1/432 subtests failed.
```
```
$ test/evp_test.exe test/evptests.txt 
line 3322: unexpected keyword PeerKey
```

dwmw2 commented:
Fairly sure *that* one isn't my fault either...
```
 $ test/evp_test.exe test/evptests.txt 
line 3322: unexpected keyword PeerKey
 $ unix2dos test/evptests.txt 
unix2dos: converting file test/evptests.txt to DOS format...
 $ test/evp_test.exe test/evptests.txt 
499 tests completed with 0 errors, 0 skipped
```

dwmw2 commented:
No changes; just a remind^H^H^H^Hbase.

mattcaswell commented on doc/ssl/DTLS_get_data_mtu.pod:
```diff
@@ -0,0 +1,36 @@
+
+This function obtains the maximum data payload size for the established
+DTLS connection B<ssl>, based on the DTLS record MTU and the overhead
+of the DTLS reord header, encryption and authentication currently in use.
```
typo: record

mattcaswell commented on doc/ssl/DTLS_get_data_mtu.pod:
```diff
@@ -0,0 +1,36 @@
+=pod
```
The location for this file needs to change now that we have restructured the doc directory

mattcaswell commented:
Aside from the above mentioned minor issues I'm about ready to plus one this.

dwmw2 commented:
On a plane now; will fix when I land. Thanks.

mattcaswell commented:
Needs a second team review

mattcaswell commented:
Merged. Thanks.

dwmw2 commented:
Thanks. Do I need to do anything to accommodate he TLSv1.3 support which is starting to go in?

mattcaswell commented:
No. TLSv1.3 only - doesn't impact DTLS.

dwmw2 commented:
I was assuming DTLSv1.3 and some actual cipher support wouldn't be far behind. Just wondering if there a way that I can extend the tests so that when DTLSv1.3 *does* come along, it can automatically be tested.
